clean_old_aggregates.py

```diff
@@ -0,0 +1,70 @@
+import keystoneauth1
+import datetime
+from keystoneauth1.identity import v3
+from keystoneauth1 import loading
+from keystoneauth1 import session
+from novaclient import client as nova_client
+from blazarclient import client as blazar_client
+
+def main():
+
+    loader = loading.get_plugin_loader('password')
+    auth = loader.load_from_options(auth_url='http://REDACTED:35357',
+    username='admin',
+    password='REDACTED',
+    project_id='REDACTED',
+    user_domain_id ='default')
+
+    sess = session.Session(auth=auth)
+
+    nova = nova_client.Client(version='2', region_name='CHI@TACC', session=sess)
+    blazar = blazar_client.Client(1, session=sess, service_type='reservation', region_name='CHI@TACC')
+
+    term_leases=[]
+    agglist=[]
+    aggdetails=[]
+    aggnames=[]
+    exp_aggs=[]
+    agg_host=[]
+
+    # Find leases with end date before current time
+    leases=blazar.lease.list()
+
+    time=datetime.datetime.now()
+    for x in leases:
+        lease_time=datetime.datetime.strptime(x['end_date'], '%Y-%m-%dT%H:%M:%S.%f')
+        if lease_time < time:
+            print("lease ending {} has ended. Time {}".format(lease_time,time))
+            term_leases.append(x['reservations'][0]['id'])
+
+    # Find aggregates matching ended leases
+    aggregates=nova.aggregates.list()
+
+    for agg in aggregates:
+        ags=str(agg)
+        agglist.append(ags[12:17])
+
+    for x in agglist:
+        aggdetails.append(nova.aggregates.get_details(x).__dict__)
+
+    for x in aggdetails:
+        aggnames.append({x['id'] : x['name']})
+
+    for term_lease_item in term_leases:
```

Comment:
So, here we're trying to do the linking of leases to aggregates. I propose to use a list comprehension again:

```python
def aggregate_for_lease(lease):
  reservation_id = lease['reservations'][0]['id']
  if not reservation_id:
    raise ValueError('Could not get reservation id from lease.')
  return next(x for x in aggregates if x.name.endswith(reservation_id))

matching_aggregates = filter(None, [aggregate_for_lease(lease) for lease in term_leases])

Reply:
FYI, the `reservation_id = lease['reservations'][0]['id']` is also a bit brittle, as it assumes that (a) there is exactly one reservation under the lease and (b) that reservation is a physical host reservation. Neither of these assumptions holds for all cases.

A more robust solution might be:

```python
# Note the plural in the function name! We are now returning a list.
def aggregates_for_lease(lease):
  physical_reservation_ids = [r['id'] for r in lease['reservations'] if r['type'] == 'physical:host']
  # Splitting aggregate name by ":" and taking the last bit should give us the reservation id.
  # (Yes, you can index by a negative number in Python.)
  return [x for x in aggregates if x.name.split(':')[-1] in physical_reservation_ids]